docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,822 @@
+    parameters with the buffers is validated.
+
+    If the field is not 3-D, as indicated by :code:`axes`, the length of parameters
+    :code:`aligned_index` and :code:`shape`, may either be of length 3 or of the actual dimension
+    of the storage, where the not needed entries are ignored in the latter case.
+
+
+The definitions of the common parameters accepted by all the previous functions are the following:
+
+:code:`dtype: dtype_like`
+    The dtype of the storage (NumPy dtype or accepted by :code:`np.dtype()`). It defaults to
+    :code:`np.float64`.
+
+:code:`shape: Sequence[int]`
+    Sequence of length :code:`ndim` (:code:`ndim` = number of dimensions, maximum 3) with the shape
+    of the storage, that is, the full addressable space in the allocated memory buffer. (See also
+    Section :ref:`domain_and_halo`)
+
+Additionally, these **optional** keyword-only parameters are accepted:
+
+:code:`aligned_index: Sequence[int]`
+    The point to which the memory is aligned, defaults to the lower indices of the halo attribute.
+
+:code:`alignment: Optional[int]`
+    Indicates on a boundary of how many elements the point :code:`aligned_index` is aligned. It
+    defaults to :code:`1`, which indicates no alignment.
+
+:code:`axes: str`
+    Any permutation of a sub-sequence of the :code:`"IJK"` string indicating the spatial dimensions
+    along which the field extends and their order for indexing operations in Python. The default
+    value is :code:`"IJK"`.
+
+:code:`defaults: Optional[str]`
+    It can be used in the way of the current :code:`backend` parameter. For each backend, as well
+    as for the keys :code:`"F"` and :code:`"C"` (equivalent to the same values in the :code:`order`
+    parameter for NumPy allocation routines) a preset of suitable parameters is provided. Explicit
+    definitions of additional parameters are possible and they override its default value from the
+    preset.
+
+:code:`device: Optional[str]`
+    Indicates whether the storage should contain a buffer on an accelerator device. Currently it
+    only accepts :code:`"gpu"` or :code:`None`. Defaults to :code:`None`.
+
+:code:`halo: Optional[Sequence[Union[int, Tuple[int, int]]]`
+    Sequence of length :code:`ndim` where each entry is either an :code:`int` or a 2-tuple
+    of :code:`int` s. A sequence of integer numbers represent a symmetric halo with the specific
+    size per dimension, while a sequence of 2-tuple specifies the start and end boundary sizes on
+    the respective dimension. It defaults to no halo, i.e. :code:`(0, 0, 0)`. (See also Section
+    :ref:`domain_and_halo`)
+
+:code:`layout: Optional[str]`
+    A length-3 string with the name of axes. The sequence indicates the order of strides in
+    decreasing order, i.e. the first entry in the sequence corresponds to the axis with the largest
+    stride. The layout map is always of length 3. The default value is "IJK".
```

Comment:
I assume you mean that storages are initialized in fortran in [k, j, i]. then you would do as_storage with `defaults="F"`, axes="KJI`. If you later permute this just like you propose, you will get an `IJK` storage with the same layout, i.e there is no copy and the smalles stride remains in `K`. This will then be valid, provided the stencil is being compiled for an `IJK` layout. (The python backends shall accept any layout, and specifying a custom layout for compiled backends was already implemented in the implementation branch but removed last-minute to clearly separate storage from DSL in the GDP. )

So yes, that will work.

Reply:
What happens is I initialize an array in Python with C-ordering and [k, j, i] dimensions. This is passed to a Fortran wrapper which copies in data to the Python-allocated array. I then want to use this array in a stencil which has been written using the first, second, and third indexes for i, j, and k.

Put another way, if I initialize a storage with KJI order or IJK order, does it make a difference what result I will get from a stencil? Or do I need to do some kind of transpose (ideally non-copying) to get the right result, and if so what would I do?